tower-http/src/validate_request.rs

```diff
@@ -0,0 +1,499 @@
+{
+    type Output = F::Output;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        match self.project().kind.project() {
+            KindProj::Future { future } => future.poll(cx),
+            KindProj::Error { response } => {
+                let response = response.take().unwrap();
+                Poll::Ready(Ok(response))
+            }
+        }
+    }
+}
+
+/// Trait for validating requests.
+pub trait ValidateRequest<B> {
+    /// The body type used for responses to unvalidated requests.
+    type ResponseBody;
+
+    /// Validate the request.
+    ///
+    /// If `Ok(())` is returned then the request is allowed through, otherwise not.
+    fn validate(&mut self, request: &mut Request<B>) -> Result<(), Response<Self::ResponseBody>>;
+}
+
+impl<B, F, ResBody> ValidateRequest<B> for F
+where
+    F: FnMut(&mut Request<B>) -> Result<(), Response<ResBody>>,
+{
+    type ResponseBody = ResBody;
+
+    fn validate(&mut self, request: &mut Request<B>) -> Result<(), Response<Self::ResponseBody>> {
+        self(request)
+    }
+}
+
+/// Type that performs validation of the Accept header.
+pub struct AcceptHeader<ResBody> {
+    header_value: HeaderValue,
+    _ty: PhantomData<fn() -> ResBody>,
+}
+
+impl<ResBody> AcceptHeader<ResBody> {
+    fn new(header_value: &str) -> Self
+    where
+        ResBody: Body + Default,
+    {
+        Self {
+            header_value: header_value
+                .parse()
+                .expect("token is not a valid header value"),
+            _ty: PhantomData,
+        }
+    }
+}
+
+impl<ResBody> Clone for AcceptHeader<ResBody> {
+    fn clone(&self) -> Self {
+        Self {
+            header_value: self.header_value.clone(),
+            _ty: PhantomData,
+        }
+    }
+}
+
+impl<ResBody> fmt::Debug for AcceptHeader<ResBody> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("AcceptHeader")
+            .field("header_value", &self.header_value)
+            .finish()
+    }
+}
+
+impl<B, ResBody> ValidateRequest<B> for AcceptHeader<ResBody>
+where
+    ResBody: Body + Default,
+{
+    type ResponseBody = ResBody;
+
+    fn validate(&mut self, req: &mut Request<B>) -> Result<(), Response<Self::ResponseBody>> {
+        if !req.headers().contains_key(header::ACCEPT) {
+            return Ok(());
+        }
+        if req
+            .headers()
+            .get_all(header::ACCEPT)
+            .into_iter()
+            .flat_map(|header| {
+                header
+                    .to_str()
+                    .ok()
+                    .into_iter()
+                    .flat_map(|s| s.split(",").map(|typ| typ.trim()))
+            })
+            .any(|h| {
+                let value = self.header_value.to_str().unwrap();
+                let primary = format!("{}/*", value.split("/").nth(0).unwrap());
+                h == "*/*" || h == primary || h == value
+            })
```

Comment:
Maybe doing the validation with the `mime` crate is gonna be more robust. wdyt?

Reply:
Also I'd really like us not to panic within a service. Currently you can provide a valid `str` that doesn't match `*/*`, `type/*` or `type/subtype` and panic in the service when that check can be done when creating the `AcceptHeader`.

Reply:
I've moved the line that can panic in `new` and used `mime`, thanks for the suggestion